modin/pandas/test/test_groupby.py

```diff
@@ -1714,3 +1766,110 @@ def test_not_str_by(by, as_index):
+        pytest.param(
+            {"b": "mean", "c": "sum", "d": "nunique"},
+            id="dict_agg_has_intersection_with_by",
+        ),
+        pytest.param(
+            {"a": "nunique", "c": "sum", "d": "nunique"},
+            id="dict_agg_has_intersection_with_categorical_by",
+        ),
+    ],
+)
+def test_handle_as_index(
+    internal_by_length, external_by_length, has_categorical_by, agg_func, request
+):
+    """
+    Test ``GroupBy.handle_as_index``.
+
+    The role of the ``handle_as_index`` method is to build a groupby result considering
+    ``as_index=False`` from the result that was computed with ``as_index=True``.
+
+    So the testing flow is the following:
+        1. Compute GroupBy result with the ``as_index=True`` parameter via Modin.
+        2. Build ``as_index=False`` result from the ``as_index=True`` using ``handle_as_index`` method.
+        3. Compute GroupBy result with the ``as_index=False`` parameter via Pandas (is the reference result).
+        4. Compare the result from the second step with the reference.
+    """
+    by_length = sum((internal_by_length, external_by_length))
+    if by_length == 0:
+        return
+
+    if (
+        has_categorical_by
+        and by_length > 1
+        and (
+            isinstance(agg_func, dict)
+            or ("nunique" in request.node.callspec.id.split("-"))
+        )
+    ):
+        pytest.skip(
+            "The linked bug makes pandas raise an exception when 'by' is categorical: "
+            "https://github.com/pandas-dev/pandas/issues/36698"
+        )
+
+    data = {
+        "a": [1, 1, 2, 2],
+        "b": [2, 2, 3, 3],
+        "c": [1, 2, 3, 4],
+        "d": [5, 6, 7, 8],
+    }
+
+    if isinstance(agg_func, dict):
+        selection = list(agg_func.keys())
+        agg_dict = agg_func
+        agg_func = lambda grp: grp.agg(agg_dict)  # noqa: E731 (lambda assignment)
+    else:
+        selection = None
+
+    df = pandas.DataFrame(data)
+    external_by_cols = GroupBy.validate_by(df.add_prefix("external_"))
+
+    if has_categorical_by:
+        df = df.astype({"a": "category"})
+
+    internal_by = df.columns[:internal_by_length].tolist()
+    external_by = external_by_cols[:external_by_length]
+
+    by = internal_by + external_by
+
+    grp_res = pd.DataFrame(df).groupby(by, as_index=True)
+    grp_ref = df.groupby(by, as_index=False)
+
+    res = agg_func(grp_res)
+    ref = agg_func(grp_ref)
+
+    drop, lvls_to_drop, cols_to_drop = GroupBy.handle_as_index(
+        result_cols=res.columns,
+        result_index_names=res.index.names,
+        internal_by_cols=internal_by,
+        by_cols_dtypes=df[internal_by].dtypes.values,
+        is_multi_by=len(by) > 1,
+        selection=selection,
+        drop=len(internal_by) != 0,
+    )
+
+    if len(lvls_to_drop) > 0:
+        res.index = res.index.droplevel(lvls_to_drop)
+    if len(cols_to_drop) > 0:
+        res = res.drop(columns=cols_to_drop)
+    res = res.reset_index(drop=drop)
```

Comment:
I saw this pattern at least 4 or 5 times already... I feel that it should be a helper function or something.

Reply:
wrapped it into the `handle_as_index_for_dataframe` function